clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -149,41 +149,57 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
      *
      * @param exception The exception to use to complete the future and awaiting callbacks.
      */
-    public void abort(RuntimeException exception) {
+    public boolean abort(RuntimeException exception) {
         if (!finalState.compareAndSet(null, FinalState.ABORTED))
-            throw new IllegalStateException("Batch has already been completed in final state " + finalState.get());
+            //throw new IllegalStateException("Batch has already been completed in final state " + finalState.get());
```

Comment:
As far as I can tell, it shouldn't be possible to abort a batch after it has been completed. Is this correct? If so, I think it might be better to continue to raise `IllegalStateException`. It's preferable to keep the allowable state transitions as narrowly defined as possible since it ensures faster failure for unexpected paths.